diff --git a/src/gallium/targets/libgl-xlib/meson.build b/src/gallium/targets/libgl-xlib/meson.build
--- a/src/gallium/targets/libgl-xlib/meson.build
+++ b/src/gallium/targets/libgl-xlib/meson.build
@@ -25,7 +25,7 @@
   link_depends : gallium_xlib_link_depends,
   link_whole : [libxlib, libglapi_bridge],
   link_with : [libgalliumvl_stub, libws_xlib, libglapi, libgallium, libmesa],
-  dependencies : [dep_x11, idep_mesautil, dep_thread, dep_clock, dep_unwind, driver_swrast, driver_virgl, driver_asahi],
+  dependencies : [dep_x11, idep_mesautil, dep_thread, dep_clock, dep_unwind, driver_swrast, driver_virgl, driver_asahi, driver_zink],
   install : true,
   version : '1.5.0',
   darwin_versions: '4.0.0',
