diff --git a/src/gallium/drivers/zink/zink_screen.c b/src/gallium/drivers/zink/zink_screen.c
--- a/src/gallium/drivers/zink/zink_screen.c
+++ b/src/gallium/drivers/zink/zink_screen.c
@@ -3189,7 +3189,9 @@ init_driver_workarounds(struct zink_screen *screen)
    if (zink_debug & ZINK_DEBUG_NOGENERAL)
       screen->driver_workarounds.general_layout = false;
 
-   if (!screen->info.have_EXT_vertex_input_dynamic_state || !screen->info.have_EXT_transform_feedback)
+   if (zink_driverid(screen) == VK_DRIVER_ID_QUALCOMM_PROPRIETARY ||
+       !screen->info.have_EXT_vertex_input_dynamic_state ||
+       !screen->info.have_EXT_transform_feedback)
       screen->info.have_KHR_device_address_commands = false;
 
    if (!screen->resizable_bar)
