diff --git a/src/gallium/drivers/zink/zink_blit.c b/src/gallium/drivers/zink/zink_blit.c
--- a/src/gallium/drivers/zink/zink_blit.c
+++ b/src/gallium/drivers/zink/zink_blit.c
@@ -80,11 +80,7 @@ blit_resolve(struct zink_context *ctx, const struct pipe_blit_info *info, bool *
       zink_resource_image_transfer_dst_barrier(ctx, dst, info->dst.level, &info->dst.box, false);
       screen->image_barrier(ctx, use_src,
                               VK_IMAGE_LAYOUT_GENERAL,
-                              VK_ACCESS_TRANSFER_READ_BIT,
-                              VK_PIPELINE_STAGE_TRANSFER_BIT);
-      screen->image_barrier(ctx, use_src,
-                              VK_IMAGE_LAYOUT_GENERAL,
-                              VK_ACCESS_TRANSFER_WRITE_BIT,
+                              VK_ACCESS_TRANSFER_READ_BIT | VK_ACCESS_TRANSFER_WRITE_BIT,
                               VK_PIPELINE_STAGE_TRANSFER_BIT);
    } else {
       zink_resource_setup_transfer_layouts(ctx, use_src, dst);
@@ -317,11 +313,7 @@ blit_native(struct zink_context *ctx, const struct pipe_blit_info *info, bool *n
       zink_resource_image_transfer_dst_barrier(ctx, dst, info->dst.level, &info->dst.box, false);
       screen->image_barrier(ctx, use_src,
                               VK_IMAGE_LAYOUT_GENERAL,
-                              VK_ACCESS_TRANSFER_READ_BIT,
-                              VK_PIPELINE_STAGE_TRANSFER_BIT);
-      screen->image_barrier(ctx, use_src,
-                              VK_IMAGE_LAYOUT_GENERAL,
-                              VK_ACCESS_TRANSFER_WRITE_BIT,
+                              VK_ACCESS_TRANSFER_READ_BIT | VK_ACCESS_TRANSFER_WRITE_BIT,
                               VK_PIPELINE_STAGE_TRANSFER_BIT);
    } else {
       zink_resource_setup_transfer_layouts(ctx, use_src, dst);
@@ -694,8 +686,7 @@ zink_blit_barriers(struct zink_context *ctx, struct zink_resource *src, struct z
                              VK_IMAGE_LAYOUT_GENERAL;
       /* apply read barrier first to avoid "sticky" read+write access flags in resource_needs_barrier() */
       screen->image_barrier(ctx, src, layout, VK_ACCESS_SHADER_READ_BIT, VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT);
-      u_foreach_bit(f, flags)
-         screen->image_barrier(ctx, src, layout, BITFIELD_BIT(f), pipeline);
+      screen->image_barrier(ctx, src, layout, flags, pipeline);
    } else {
       if (src) {
          VkImageLayout layout = screen->driver_workarounds.general_layout ? VK_IMAGE_LAYOUT_GENERAL :
@@ -711,8 +702,7 @@ zink_blit_barriers(struct zink_context *ctx, struct zink_resource *src, struct z
                              util_format_is_depth_or_stencil(dst->base.b.format) ?
                              VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL :
                              VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL;
-      u_foreach_bit(f, flags)
-         screen->image_barrier(ctx, dst, layout, BITFIELD_BIT(f), pipeline);
+      screen->image_barrier(ctx, dst, layout, flags, pipeline);
    }
    if (!ctx->unordered_blitting)
       zink_resource_disable_unordered(dst, true);
diff --git a/src/gallium/drivers/zink/zink_context.c b/src/gallium/drivers/zink/zink_context.c
--- a/src/gallium/drivers/zink/zink_context.c
+++ b/src/gallium/drivers/zink/zink_context.c
@@ -2871,8 +2871,7 @@ zink_set_global_binding(struct pipe_context *pctx,
          addr += zink_resource_get_address(zink_screen(pctx->screen), res);
          memcpy(handles[i], &addr, sizeof(addr));
          zink_resource_usage_set(res, ctx->bs, true);
-         zink_screen(ctx->base.screen)->buffer_barrier(ctx, res, VK_ACCESS_SHADER_READ_BIT, VK_PIPELINE_STAGE_COMPUTE_SHADER_BIT);
-         zink_screen(ctx->base.screen)->buffer_barrier(ctx, res, VK_ACCESS_SHADER_WRITE_BIT, VK_PIPELINE_STAGE_COMPUTE_SHADER_BIT);
+         zink_screen(ctx->base.screen)->buffer_barrier(ctx, res, VK_ACCESS_SHADER_READ_BIT | VK_ACCESS_SHADER_WRITE_BIT, VK_PIPELINE_STAGE_COMPUTE_SHADER_BIT);
          zink_resource_disable_unordered(res, true);
       } else if (globals[i]) {
          zink_batch_reference_resource(ctx, zink_resource(globals[first + i]));
@@ -3116,8 +3115,7 @@ prep_fb_attachment(struct zink_context *ctx, struct zink_resource *res, unsigned
    else if (layout != VK_IMAGE_LAYOUT_ATTACHMENT_FEEDBACK_LOOP_OPTIMAL_EXT &&
             i >= ctx->fb_state.nr_cbufs && screen->driver_workarounds.general_depth_layout)
       layout = VK_IMAGE_LAYOUT_GENERAL;
-   u_foreach_bit(a, access)
-      screen->image_barrier(ctx, res, layout, BITFIELD_BIT(a), pipeline);
+   screen->image_barrier(ctx, res, layout, access, pipeline);
    if (!(res->aspect & VK_IMAGE_ASPECT_COLOR_BIT))
       ctx->zsbuf_readonly = res->layout == VK_IMAGE_LAYOUT_DEPTH_STENCIL_READ_ONLY_OPTIMAL;
    zink_resource_disable_unordered(res, true);
@@ -3755,11 +3753,7 @@ zink_flush_clears(struct zink_context *ctx)
          }
          screen->image_barrier(ctx, res,
                               general_layout ? VK_IMAGE_LAYOUT_GENERAL : VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL,
-                              VK_ACCESS_COLOR_ATTACHMENT_READ_BIT,
-                              VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT);
-         screen->image_barrier(ctx, res,
-                              general_layout ? VK_IMAGE_LAYOUT_GENERAL : VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL,
-                              VK_ACCESS_COLOR_ATTACHMENT_WRITE_BIT,
+                              VK_ACCESS_COLOR_ATTACHMENT_READ_BIT | VK_ACCESS_COLOR_ATTACHMENT_WRITE_BIT,
                               VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT);
          zink_resource_disable_unordered(res, true);
          assert(res->layout != VK_IMAGE_LAYOUT_UNDEFINED);
@@ -3768,11 +3762,7 @@ zink_flush_clears(struct zink_context *ctx)
          struct zink_resource *res = zink_resource(ctx->fb_state.zsbuf.texture);
          screen->image_barrier(ctx, res,
                               general_layout ? VK_IMAGE_LAYOUT_GENERAL : VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL,
-                              VK_ACCESS_DEPTH_STENCIL_ATTACHMENT_READ_BIT,
-                              VK_PIPELINE_STAGE_EARLY_FRAGMENT_TESTS_BIT | VK_PIPELINE_STAGE_LATE_FRAGMENT_TESTS_BIT);
-         screen->image_barrier(ctx, res,
-                              general_layout ? VK_IMAGE_LAYOUT_GENERAL : VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL,
-                              VK_ACCESS_DEPTH_STENCIL_ATTACHMENT_WRITE_BIT,
+                              VK_ACCESS_DEPTH_STENCIL_ATTACHMENT_READ_BIT | VK_ACCESS_DEPTH_STENCIL_ATTACHMENT_WRITE_BIT,
                               VK_PIPELINE_STAGE_EARLY_FRAGMENT_TESTS_BIT | VK_PIPELINE_STAGE_LATE_FRAGMENT_TESTS_BIT);
          zink_resource_disable_unordered(res, true);
          assert(res->layout != VK_IMAGE_LAYOUT_UNDEFINED);
@@ -5479,11 +5469,7 @@ zink_resource_copy_region(struct pipe_context *pctx,
          zink_resource_image_transfer_dst_barrier(ctx, dst, dst_level, &box, false);
          screen->image_barrier(ctx, src,
                                VK_IMAGE_LAYOUT_GENERAL,
-                               VK_ACCESS_TRANSFER_READ_BIT,
-                               VK_PIPELINE_STAGE_TRANSFER_BIT);
-         screen->image_barrier(ctx, src,
-                               VK_IMAGE_LAYOUT_GENERAL,
-                               VK_ACCESS_TRANSFER_WRITE_BIT,
+                               VK_ACCESS_TRANSFER_READ_BIT | VK_ACCESS_TRANSFER_WRITE_BIT,
                                VK_PIPELINE_STAGE_TRANSFER_BIT);
       } else {
          zink_resource_setup_transfer_layouts(ctx, src, dst);
diff --git a/src/gallium/drivers/zink/zink_resource.c b/src/gallium/drivers/zink/zink_resource.c
--- a/src/gallium/drivers/zink/zink_resource.c
+++ b/src/gallium/drivers/zink/zink_resource.c
@@ -3502,11 +3502,7 @@ zink_resource_setup_transfer_layouts(struct zink_context *ctx, struct zink_resou
        */
       screen->image_barrier(ctx, src,
                             VK_IMAGE_LAYOUT_GENERAL,
-                            VK_ACCESS_TRANSFER_READ_BIT,
-                            VK_PIPELINE_STAGE_TRANSFER_BIT);
-      screen->image_barrier(ctx, src,
-                            VK_IMAGE_LAYOUT_GENERAL,
-                            VK_ACCESS_TRANSFER_WRITE_BIT,
+                            VK_ACCESS_TRANSFER_READ_BIT | VK_ACCESS_TRANSFER_WRITE_BIT,
                             VK_PIPELINE_STAGE_TRANSFER_BIT);
    } else {
       screen->image_barrier(ctx, src,
